src/core/storage/fileio/s3_api.cpp

```diff
@@ -111,20 +114,57 @@ bool bucket_name_valid(const std::string& bucket_name) {
 }
 
 
-std::string string_from_s3url(const s3url& parsed_url) {
-  std::string ret = "s3://" + parsed_url.access_key_id + ":"
-      + parsed_url.secret_key + ":";
+} // anonymous namespace
+
+S3Client init_aws_sdk_with_turi_env(s3url& parsed_url) {
+  // s3 client config
+  // DefaultCredentialProviderChain
+  Aws::Client::ClientConfiguration clientConfiguration;
+
+  // a little bit too long, anyway
+  clientConfiguration.requestTimeoutMs = 5 * 60000;
+  clientConfiguration.connectTimeoutMs = 20000;
+
+  if (turi::fileio::insecure_ssl_cert_checks()) {
+    clientConfiguration.verifySSL = false;
+  }
+
   if (!parsed_url.endpoint.empty()) {
-    ret += parsed_url.endpoint + "/";
+    clientConfiguration.endpointOverride = parsed_url.endpoint.c_str();
+  } else {
+    auto env_var = get_endpoint_from_env();
+    if (env_var) {
+      clientConfiguration.endpointOverride = env_var->c_str();
+      parsed_url.sdk_endpoint = env_var->c_str();
+    }
   }
-  ret += parsed_url.bucket;
-  if (!parsed_url.object_name.empty()) {
-    ret += "/" + parsed_url.object_name;
+
+  // TODO: add proxy support
+  // clientConfiguration.proxyHost = proxy.c_str();
+
+  std::string region = fileio::get_region_name_from_endpoint(
+      clientConfiguration.endpointOverride.c_str());
+
+  if (!region.empty()) {
+    clientConfiguration.region = region.c_str();
+  } else {
+    auto env_var = get_auth_region_from_env();
+    if (env_var) {
+      clientConfiguration.region = env_var->c_str();
```

Comment:
The pointers returned from c_str() become invalid after env_var goes out of scope and may cause a segfault unless the .region and .sdk_region perform a copy on assignment.  I think we should put these strings (or optionsals) in the parsed_url struct to make sure the data is always valid.  I.e. add an std::string env_auth_region (may be empty) and a std::string env_endpoint; field to the parsed_url and then set these clientConfigurations from that.

Reply:
Otherwise, just make sure they're at function scope.

Reply:
clientConfiguration.region  is a Aws::string, which cannot copy the std::string. Using `c_str` is to force it copy the contents from std::string.

Reply:
use `c_str` is also a very common when using reference-count-based std::string for the multithreaded environment.

Reply:
Oh, then it does copy it.  So we're okay.  

Yes, it is common, but the original std::string cannot be deleted while the .c_str() result is being used.

Reply:
It will be deleted in recent std::string implementation since it returns a const char*. There's no way to track a pointer usage. You can easily get segfault from dangling c_str() pointer in current impl.